src/idns/output/dig.rs

```diff
@@ -0,0 +1,165 @@
+                    Expire(expire) => {
+                        writeln!(target, "; EXPIRE: {}", expire)?
+                    }
+                    TcpKeepalive(opt) => {
+                        writeln!(target, "; TCPKEEPALIVE: {}", opt)?
+                    }
+                    Padding(padding) => {
+                        writeln!(target, "; PADDING: {}", padding)?
+                    }
+                    ClientSubnet(opt) => {
+                        writeln!(target, "; CLIENTSUBNET: {}", opt)?
+                    }
+                    Cookie(cookie) => {
+                        writeln!(target, "; COOKIE: {}", cookie)?
+                    }
+                    Chain(chain) => {
+                        writeln!(target, "; CHAIN: {}", chain)?
+                    }
+                    KeyTag(keytag) => {
+                        writeln!(target, "; KEYTAG: {}", keytag)?
+                    }
+                    ExtendedError(extendederror) => {
+                        writeln!(target, "; EDE: {}", extendederror)?
+                    }
+                    Other(other) => {
+                        writeln!(target, "; {}", other.code())?;
+                    }
+                    _ => writeln!(target, "Unknown OPT")?,
+                }
+                Err(err) => {
+                    writeln!(target, "; ERROR: bad option: {}.", err)?;
+                }
+            }
+        }
+    }
+
+    // Question
+    let questions = msg.question();
+    if counts.qdcount() > 0 {
+        write!(target, ";; QUESTION SECTION:")?;
+        for item in questions {
+            match item {
+                Ok(item) => writeln!(target, "; {}", item)?,
+                Err(err) => {
+                    writeln!(target, "; ERROR: bad question: {}.", err)?;
+                    return Ok(())
+                }
+            }
+        }
+    }
+    
+    /* Answer */
+    let mut section = match questions.answer() {
+        Ok(section) => section.limit_to::<AllRecordData<_, _>>(),
+        Err(err) => {
+            writeln!(target, "; ERROR: bad question: {}.", err)?;
+            return Ok(())
+        }
+    };
+    if counts.ancount() > 0 {
+        writeln!(target, "\n;; ANSWER SECTION:")?;
+        while let Some(item) = section.next() {
+            match item {
+                Ok(item) => writeln!(target, "{}", item)?,
+                Err(err) => {
+                    writeln!(target, "; Error: bad record: {}.", err)?;
+                    return Ok(())
+                }
+            }
+        }
+    }
+
+    // Authority
+    let mut section = match section.next_section() {
+        Ok(section) => section.unwrap().limit_to::<AllRecordData<_, _>>(),
+        Err(err) => {
+            writeln!(target, "; ERROR: bad record: {}.", err)?;
+            return Ok(())
+        }
+    };
+    if counts.nscount() > 0 {
+        writeln!(target, "\n;; AUTHORITY SECTION:")?;
+        while let Some(item) = section.next() {
```

Comment:
I have to admit I'm also a bit confused by the `next_section` API. It feels very iterator like, but like it's represented as a linked list?

Reply:
> I agree with the clippy lints here

TIL about `Iterator::by_ref` \o/

Reply:
Re `next_section`: A DNS message has three record sections that are all the same, so we have on type, `RecordSection<..>` for them. They are arranged sequentially in the message, you only know the number of records for each, and each record has a different length . Thus, you need to iterate over all the records to get to the next section. Because we do lazy parsing, `next_section` errors out if it fails to parse a record. It returns an option because when you are on the last section, there is no next section.

I want to change this by giving `RecordSection<..>` another type argument that is a sentinel so that instead of `next_section` you get, e.g., a method `authority` for `RecordSection<'_, AnswerSection, _>` that proceeds to the authority section without an option. Just hasn’t happened yet.